components/eventing-controller/pkg/backend/jetstreamv2/jetstream.go

```diff
@@ -387,81 +389,112 @@ func (js *JetStream) deleteConsumerFromJetStream(name string) error {
+func (js *JetStream) createNATSSubscription(subscription *eventingv1alpha2.Subscription, subject eventingv1alpha2.EventType, jsSubject string, jsSubKey SubscriptionSubjectIdentifier, asyncCallback func(m *nats.Msg), log *zap.SugaredLogger) error {
+	maxInFlight := subscription.GetMaxInFlightMessages(js.namedLogger())
+	jsSubscription, err := js.jsCtx.Subscribe(
+		jsSubject,
+		asyncCallback,
+		js.getDefaultSubscriptionOptions(jsSubKey, maxInFlight)...,
+	)
 	if err != nil {
-		return err
+		return backenderrors.NewFailedToSubscribeOnNATSError(err)
 	}
-	for _, subject := range subscription.Status.Types {
-		jsSubject := js.getJetStreamSubject(subscription.Spec.Source, subject.CleanType, subscription.Spec.TypeMatching)
-		jsSubKey := NewSubscriptionSubjectIdentifier(subscription, jsSubject)
-		log := log.With("subject", subject)
+	// save created JetStream subscription in storage
+	js.subscriptions[jsSubKey] = &Subscription{Subscription: jsSubscription}
+	js.metricsCollector.RecordEventTypes(subscription.Name, subscription.Namespace, subject.CleanType, jsSubKey.ConsumerName())
+	log.Debugw("Created subscription on JetStream")
 
-		if _, ok := js.subscriptions[jsSubKey]; ok {
-			continue
-		}
+	return nil
+}
 
-		// TODO: optimize this call of ConsumerInfo
-		consumerInfo, err := js.jsCtx.ConsumerInfo(js.Config.JSStreamName, jsSubKey.ConsumerName())
-		if err != nil && err != nats.ErrConsumerNotFound {
-			log.Errorw("Failed to get consumer info", "error", err)
-			return err
-		}
+// bindConsumersForInvalidNATSSubscriptions tries to bind the invalid NATS Subscription to the existing consumer.
+func (js *JetStream) bindConsumersForInvalidNATSSubscriptions(jsSubject string, jsSubKey SubscriptionSubjectIdentifier, asyncCallback func(m *nats.Msg), log *zap.SugaredLogger) error {
+	log.Debugw("Recreating subscription on JetStream because it was invalid")
+	// bind the existing consumer to a new subscription on JetStream
+	jsSubscription, err := js.jsCtx.Subscribe(
+		jsSubject,
+		asyncCallback,
+		nats.Bind(js.Config.JSStreamName, jsSubKey.ConsumerName()),
+	)
+	if err != nil {
+		return backenderrors.NewFailedToSubscribeOnNATSError(err)
+	}
+	// save recreated JetStream subscription in storage
+	js.subscriptions[jsSubKey] = &Subscription{Subscription: jsSubscription}
+	log.Debugw("Recreated subscription on JetStream")
+	return nil
+}
 
-		// skip the subject, in case there is a bound consumer on NATS
-		if consumerInfo != nil && consumerInfo.PushBound {
-			continue
-		}
+// checkSubscriptionConfig checks that the latest Subscription Config changes are propagated to the consumer.
+// In our case config contains only the "maxInFlightMessages" property, which is the maxAckPending on the consumer side.
+func (js *JetStream) checkSubscriptionConfig(subscription *eventingv1alpha2.Subscription, consumerInfo nats.ConsumerInfo, log *zap.SugaredLogger) error {
```

Comment:
nit: IMHO we are updating the consumer config and not only checking it. Another name could be `syncConsumerConfig`.

```suggestion
func (js *JetStream) updateConsumerConfig(subscription *eventingv1alpha2.Subscription, consumerInfo nats.ConsumerInfo, log *zap.SugaredLogger) error {
```

Reply:
renamed to `syncConsumerMaxInFlight` since it's the most suitable one.